### PR TITLE
Add support for a compile database file not at the root of a workspace

### PR DESCRIPTION
**Pull request by GeorgeLyon**
Introduce a new command-line flag (`--compile-db-search-path`) to `sourcekit-lsp`, which can be used to specify which subdirectories in a workspace should be searched for a clangd-style compile database (`compile_commands.json` or `compile_flags.txt`). Originally, I was going to add a `clangd`-like `--compile-commands-dir` flag, but upon further investigation decided this wasn't a good approach, as sourcekit-lsp supports multiple workspaces and `clangd` [does not](https://github.com/clangd/clangd/issues/1549). In a world with multiple workspaces it does not make sense to specify a single absolute path to a compile-database, so instead I opted for a set of relative search paths, which apply to all databases managed by a particular instance of `sourcekit-lsp`.

### CODE REVIEW

File: Sources/SKCore/CompilationDatabase.swift [hunk 1]
```diff
@@ -185,12 +197,18 @@ extension JSONCompilationDatabase: Codable {
 }
 
 extension JSONCompilationDatabase {
-  public init(directory: AbsolutePath, _ fileSystem: FileSystem = localFileSystem) throws {
+  public init?(directory: AbsolutePath, _ fileSystem: FileSystem = localFileSystem) throws {
```
Comment (GeorgeLyon) [inline Sources/SKCore/CompilationDatabase.swift:200]: Previously, this would throw an error both if the file was missing or if parsing failed. I updated the behavior to simply return `nil` if the file does not exist, allowing me to more correctly implement the search behavior in `tryLoadCompilationDatabase` above. This makes things slightly more strict than they were, in the unlikely case that both types of database are present _and_ the JSON fails to parse this will result in an error rather than falling back to compile_flags.txt

File: Sources/SKTestSupport/SKSwiftPMTestWorkspace.swift [hunk 2]
```diff
@@ -55,7 +55,7 @@ public final class SKSwiftPMTestWorkspace {
   /// Connection to the language server.
   public let testClient: TestSourceKitLSPClient
 
-  /// When `testServer` is not `nil`, the workspace will be opened in that server, otherwise a new server will be created for the workspace
+  /// When `testClient` is not `nil`, the workspace will be opened in that client's server, otherwise a new client will be created for the workspace
```
Comment (GeorgeLyon) [inline Sources/SKTestSupport/SKSwiftPMTestWorkspace.swift:58]: Not really related, but I was confused by the wording here when browsing the codebase.

File: Sources/sourcekit-lsp/SourceKitLSP.swift [hunk 3]
```diff
@@ -137,6 +150,18 @@ struct SourceKitLSP: ParsableCommand {
...
+  @Option(
+    name: .customLong("compilation-db-search-path", withSingleDash: false),
+    parsing: .unconditionalSingleValue,
+    help:
+      "Specify a relative path where SourceKit should search for `compile_commands.json` or `compile_flags.txt` relative to the root of a workspace. Multiple search paths may be specified by repeating this option."
+  )
+  var compilationDatabaseSearchPaths = try! [
+    // These defaults match the behavior of clangd
+    RelativePath(validating: "."),
+    RelativePath(validating: "build"),
```
Comment (GeorgeLyon) [inline Sources/sourcekit-lsp/SourceKitLSP.swift:162]: It may make sense to pull these out and make the search paths "additional" so the defaults persist even in the presence of custom search paths.

File: Sources/SKCore/CompilationDatabase.swift (hunk 1 above)
Comment (ahoppen) [reply Sources/SKCore/CompilationDatabase.swift:204]: That makes sense to me. Could you add a doc-comment to the initializer explaining when this returns `nil` and when it throws?

File: Sources/SKCore/CompilationDatabase.swift [hunk 4]
```diff
@@ -70,11 +71,19 @@ public protocol CompilationDatabase {
...
-  return
-    (try? JSONCompilationDatabase(directory: directory, fileSystem))
-    ?? (try? FixedCompilationDatabase(directory: directory, fileSystem))
+  return searchPaths
+    .lazy
+    .map { directory.appending($0) }
+    .compactMap {
+      try?
+        (JSONCompilationDatabase(directory: $0, fileSystem)
+        ?? FixedCompilationDatabase(directory: $0, fileSystem))
```
Comment (ahoppen) [inline Sources/SKCore/CompilationDatabase.swift:83]: If the initializers now return `nil` when the file doesn’t exist, I think we no longer need to do `try?` here and can change it to `try`. That way we don’t silently ignore errors anymore.

File: Sources/SKTestSupport/SKSwiftPMTestWorkspace.swift (hunk 2 above)
Comment (ahoppen) [reply Sources/SKTestSupport/SKSwiftPMTestWorkspace.swift:58]: Oh, I recently changed `testServer` to `testClient` and looks like I forgot that comment. Thanks for updating it.

File: Sources/sourcekit-lsp/SourceKitLSP.swift [hunk 5]
```diff
@@ -137,6 +150,18 @@ struct SourceKitLSP: ParsableCommand {
   )
   var indexPrefixMappings = [PathPrefixMapping]()
 
+  @Option(
+    name: .customLong("compilation-db-search-path", withSingleDash: false),
+    parsing: .unconditionalSingleValue,
```
Comment (ahoppen) [inline Sources/sourcekit-lsp/SourceKitLSP.swift:155]: Any reason why you decided to choose `unconditionalSingleValue` here instead of the default `singleValue`?

File: Tests/SourceKitLSPTests/CompilationDatabaseTests.swift [hunk 6]
```diff
@@ -43,7 +43,9 @@ final class CompilationDatabaseTests: XCTestCase {
     let compilationDatabaseUrl = ws.builder.buildRoot.appendingPathComponent("compile_commands.json")
 
     _ = try ws.sources.edit({ builder in
-      let compilationDatabase = try JSONCompilationDatabase(file: AbsolutePath(validating: compilationDatabaseUrl.path))
+      let compilationDatabase = try JSONCompilationDatabase(
+        file: AbsolutePath(validating: compilationDatabaseUrl.path)
+      )!
       let newCommands = compilationDatabase.allCommands.map {
```
Comment (ahoppen) [inline Tests/SourceKitLSPTests/CompilationDatabaseTests.swift:49]: Could you use `XCTUnwrap` instead of `!` here so the test doesn’t crash if the database is `nil`?

File: Sources/sourcekit-lsp/SourceKitLSP.swift [hunk 7]
```diff
@@ -137,6 +150,18 @@ struct SourceKitLSP: ParsableCommand {
   )
   var indexPrefixMappings = [PathPrefixMapping]()
 
+  @Option(
+    name: .customLong("compilation-db-search-path", withSingleDash: false),
```
Comment (ahoppen) [inline Sources/sourcekit-lsp/SourceKitLSP.swift:154]: `withSingleDash` defaults to `false`, so we don’t need to specify it.

File: Sources/sourcekit-lsp/SourceKitLSP.swift (hunk 3 above)
Comment (ahoppen) [reply Sources/sourcekit-lsp/SourceKitLSP.swift:162]: Yes, I think that would make sense. IMO the search precedence should be user-defined search paths, `.`, `build`.

File: Sources/sourcekit-lsp/SourceKitLSP.swift [hunk 8]
```diff
@@ -137,6 +150,18 @@ struct SourceKitLSP: ParsableCommand {
...
+    parsing: .unconditionalSingleValue,
+    help:
+      "Specify a relative path where SourceKit should search for `compile_commands.json` or `compile_flags.txt` relative to the root of a workspace. Multiple search paths may be specified by repeating this option."
```
Comment (ahoppen) [inline Sources/sourcekit-lsp/SourceKitLSP.swift:157]: Nitpick

```suggestion
      "Specify a relative path where sourcekit-lsp should search for `compile_commands.json` or `compile_flags.txt` relative to the root of a workspace. Multiple search paths may be specified by repeating this option."
```

**Review by ahoppen — COMMENTED** [review]
Nice. Thanks for opening a PR for this so quickly 🤩.

I would like to think about the naming of the command line option a little bit more. I’m not unconditionally happy with it but don’t have any better ideas either, so maybe it’s fine. @bnbarham Maybe you’ve got an opinion here as well.

File: Sources/sourcekit-lsp/SourceKitLSP.swift (hunk 5 above)
Comment (GeorgeLyon) [reply Sources/sourcekit-lsp/SourceKitLSP.swift:155]: No strong reason, though now that I think about it `singleValue` might be better as --compilation-db-search-path --some-other-flag may have a better compilation error.

**Comment by bnbarham** [comment]
> I would like to think about the naming of the command line option a little bit more. I’m not unconditionally happy with it but don’t have any better ideas either, so maybe it’s fine. @bnbarham Maybe you’ve got an opinion here as well.

No particularly strong ones. `compile-commands-search-path` maybe?

File: Sources/sourcekit-lsp/SourceKitLSP.swift (hunk 3 above)
Comment (GeorgeLyon) [reply Sources/sourcekit-lsp/SourceKitLSP.swift]: Done.

File: Sources/SKCore/CompilationDatabase.swift (hunk 4 above)
Comment (GeorgeLyon) [reply Sources/SKCore/CompilationDatabase.swift]: I looked into this but it is actually a bigger change than it seems as the `throws` will propagate all the way up to things like the `BuildSystem` protocol, so you may want to do that separately (if at all). In the meantime I made it `try!` since decoding errors should be very infrequent and shouldn't fail silently.

**Comment by GeorgeLyon** [comment]
> No particularly strong ones. compile-commands-search-path maybe?

This confused me when I was looking into this. I was used to "compile_commands.json", but the term "compile database" is more accurate (and what is used in the `sourcekit-lsp` code itself), since you could use "compile_flags.txt" as well. If we were being 100% pedantic (and we are all Swift developers, no? 😜) it would be something like "clang-style-compile-database-search-path", but even that may not be 100% correct. I think the current form is a good balance between being specific and being correct and not being too wordy, and the help text is pretty explicit which helps.

File: Sources/SKCore/CompilationDatabase.swift [hunk 9]
```diff
@@ -67,14 +68,28 @@ public protocol CompilationDatabase {
...
   directory: AbsolutePath,
+  additionalSearchPaths: [RelativePath] = [],
   _ fileSystem: FileSystem = localFileSystem
 ) -> CompilationDatabase? {
+  let searchPaths =
+    try! additionalSearchPaths + [
+      // These default search paths match the behavior of `clangd`
+      RelativePath(validating: "."),
+      RelativePath(validating: "build"),
+    ]
```
Comment (ahoppen) [inline Sources/SKCore/CompilationDatabase.swift:82]: Nitpick: I would move `try!` in front of `RelativePath` just to make it clear that the array creation or accessing `additionalPaths` can’t throw. This is just me being overly careful with `try!`.

```suggestion
    additionalSearchPaths + [
      // These default search paths match the behavior of `clangd`
      try! RelativePath(validating: "."),
      try! RelativePath(validating: "build"),
    ]
```

File: Sources/SKCore/CompilationDatabase.swift [hunk 10]
```diff
@@ -99,13 +114,24 @@ public struct FixedCompilationDatabase: CompilationDatabase, Equatable {
...
-  public init(directory: AbsolutePath, _ fileSystem: FileSystem = localFileSystem) throws {
+  /// Loads the compilation database located in `directory`, if any.
+  /// returns: `nil` if `compile_flags.txt` was not found
+  public init?(directory: AbsolutePath, _ fileSystem: FileSystem = localFileSystem) throws {
     let path = directory.appending(component: "compile_flags.txt")
-    try self.init(file: path, fileSystem)
+    guard let database = try Self(file: path, fileSystem) else {
+      return nil
+    }
+    self = database
```
Comment (ahoppen) [inline Sources/SKCore/CompilationDatabase.swift:124]: Am I missing something or could this just continue being

```suggestion
    try self.init(file: path, fileSystem)
```

File: Sources/SKCore/CompilationDatabase.swift [hunk 11]
```diff
@@ -185,12 +211,22 @@ extension JSONCompilationDatabase: Codable {
 }
 
 extension JSONCompilationDatabase {
-  public init(directory: AbsolutePath, _ fileSystem: FileSystem = localFileSystem) throws {
+  /// Loads the compilation database located in `directory`, if any.
+  /// returns: `nil` if `compile_commands.json` was not found
```
Comment (ahoppen) [inline Sources/SKCore/CompilationDatabase.swift:215]: Let’s use docc markup here. Same for all the other `return:` lines

```suggestion
  ///
  /// - Returns: `nil` if `compile_commands.json` was not found
```

File: Sources/SKCore/CompilationDatabase.swift [hunk 12]
```diff
@@ -185,12 +211,22 @@ extension JSONCompilationDatabase: Codable {
...
+  public init?(directory: AbsolutePath, _ fileSystem: FileSystem = localFileSystem) throws {
     let path = directory.appending(component: "compile_commands.json")
-    try self.init(file: path, fileSystem)
+    guard let database = try Self(file: path, fileSystem) else {
+      return nil
+    }
+    self = database
```
Comment (ahoppen) [inline Sources/SKCore/CompilationDatabase.swift:221]: Also here, couldn’t this be

```suggestion
    try self.init(file: path, fileSystem)
```

File: Sources/sourcekit-lsp/SourceKitLSP.swift (hunk 5 above)
Comment (ahoppen) [reply Sources/sourcekit-lsp/SourceKitLSP.swift]: That’s what I thought as well.

**Comment by ahoppen** [comment]
> No particularly strong ones. `compile-commands-search-path` maybe?

I think `compilation-db-search-path`  is better because it also handles `compile_flags.txt`. For the lack of better ideas, let’s leave it for now.

**Review by ahoppen — COMMENTED** [review]
I have a couple more minor formatting comments inline.

File: Sources/SKCore/CompilationDatabase.swift (hunk 10 above)
Comment (GeorgeLyon) [reply Sources/SKCore/CompilationDatabase.swift]: Good catch. I've had one foot outside of Swift land the year or so and forgot we can have nice things :)

File: Sources/SKCore/CompilationDatabase.swift [hunk 13]
```diff
@@ -99,13 +114,21 @@ public struct FixedCompilationDatabase: CompilationDatabase, Equatable {
...
+  /// Loads the compilation database located in `directory`, if any.
+  /// returns: `nil` if `compile_flags.txt` was not found
+  public init?(directory: AbsolutePath, _ fileSystem: FileSystem = localFileSystem) throws {
     let path = directory.appending(component: "compile_flags.txt")
     try self.init(file: path, fileSystem)
   }
 
-  public init(file: AbsolutePath, _ fileSystem: FileSystem = localFileSystem) throws {
+  /// Loads the compilation database from `file`
+  /// returns: `nil` if the file does not exist
```
Comment (ahoppen) [inline Sources/SKCore/CompilationDatabase.swift:125]: Could you also update the other `returns:` lines to use docc markup?

**Review by ahoppen — COMMENTED** [review]
Just one more note on the doc comments. 

Also: Could you squash your commits? It makes for a nicer git history.

**Comment by GeorgeLyon** [comment]
Updated comments, and squash-force0pushed... do you guys not use GitHub's "squash and merge"?

**Comment by ahoppen** [comment]
> Updated comments, and squash-force0pushed... do you guys not use GitHub's "squash and merge"?

I prefer not to squash and merge for the reasons outlined here: https://github.com/apple/swift-syntax/blob/main/CONTRIBUTING.md#authoring-commits

**Review by ahoppen — APPROVED** [review]
Nice. Let’s get this in 🚢

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Comment by GeorgeLyon** [comment]
Thanks for working with me on this @ahoppen!

**Comment by ahoppen** [comment]
Thanks for your contribution!
